### PR TITLE
[node-manager] Errors when adding a NodeUser do not block bashible

### PR DESCRIPTION
**Pull request by aleksey-su**
## Description
Errors when adding a NodeUser do not block bashible

## Why do we need it, and what problem does it solve?

#6903

## What is the expected result?
Errors when adding a NodeUser do not block bashable

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: node-manager
type: fix
summary: Errors when adding a NodeUser do not block bashible.
impact_level: default
```

### CODE REVIEW

**Comment by aleksey-su** [comment]
image

**Comment by aleksey-su** [comment]
image

**Comment by aleksey-su** [comment]
image

**Comment by aleksey-su** [comment]
image

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @aleksey-su. I've started the workflow for label 'e2e/run/openstack'...

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 1]
```diff
@@ -23,6 +23,105 @@ if bb-flag? disruption && bb-flag? reboot; then
   exit 0
 fi
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:25]: Needs comment about why this step is 000_*

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 2]
```diff
@@ -23,6 +23,105 @@ if bb-flag? disruption && bb-flag? reboot; then
...
 
+function error_check() {
+  error_code=$?
+  (( error_code != 0 ))
+}
+
+function kubectl_exec() {
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:31]: bb-kubectl can be used here

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 3]
```diff
@@ -23,6 +23,105 @@ if bb-flag? disruption && bb-flag? reboot; then
...
+  local failure_limit=3
+
+  if type kubectl >/dev/null 2>&1 && test -f /etc/kubernetes/kubelet.conf ; then
+    until kubectl_exec "${username}" --type=json --patch="${data}" --subresource=status; do
+      failure_count=$((failure_count + 1))
+      if [[ $failure_count -eq $failure_limit ]]; then
+        >&2 echo "ERROR: Failed to patch NodeUser with kubectl --kubeconfig=/etc/kubernetes/kubelet.conf"
+        break
+      fi
+      >&2 echo "failed to NodeUser with kubectl --kubeconfig=/etc/kubernetes/kubelet.conf"
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:53]:
```suggestion
      bb-error "failed to NodeUser with kubectl --kubeconfig=/etc/kubernetes/kubelet.conf"
```

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 4]
```diff
@@ -23,6 +23,105 @@ if bb-flag? disruption && bb-flag? reboot; then
...
+      sleep 10
+    done
+  elif [ -f /var/lib/bashible/bootstrap-token ]; then
+    local patch_pending=true
+    while [ "$patch_pending" = true ] ; do
+      for server in {{ .normal.apiserverEndpoints | join " " }} ; do
+        local server_addr=$(echo $server | cut -f1 -d":")
+        until local tcp_endpoint="$(ip ro get ${server_addr} | grep -Po '(?<=src )([0-9\.]+)')"; do
+          echo "The network is not ready for connecting to apiserver yet, waiting..."
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:62]:
```suggestion
          bb-log-info "The network is not ready for connecting to apiserver yet, waiting..."
```

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 5]
```diff
@@ -23,6 +23,105 @@ if bb-flag? disruption && bb-flag? reboot; then
...
+  error_code=$?
+  (( error_code != 0 ))
+}
+
+function kubectl_exec() {
+  kubectl --request-timeout 60s --kubeconfig=/etc/kubernetes/kubelet.conf patch nodeusers.deckhouse.io "${1}" "${2}" "${3}" "${4}"
+}
+
+# $1 - username $2 - request data
+function nodeuser_patch() {
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:36]: think about move it in bashbooster. Looks like we can reuse this function in multiple locations

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 6]
```diff
@@ -23,6 +23,105 @@ if bb-flag? disruption && bb-flag? reboot; then
...
+
+          echo "Successfully patched NodeUser."
+          patch_pending=false
+
+          break
+        else
+          failure_count=$((failure_count + 1))
+
+          if [[ $failure_count -eq $failure_limit ]]; then
+            >&2 echo "Failed to patch NodeUser. Number of attempts exceeded. NodeUser patch will be skipped."
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:84]:
```suggestion
            bb-log-error "Failed to patch NodeUser. Number of attempts exceeded. NodeUser patch will be skipped."
```

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 7]
```diff
@@ -23,6 +23,105 @@ if bb-flag? disruption && bb-flag? reboot; then
   exit 0
 fi
 
+function error_check() {
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:26]: use bb-error

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 8]
```diff
@@ -23,6 +23,105 @@ if bb-flag? disruption && bb-flag? reboot; then
...
+        else
+          failure_count=$((failure_count + 1))
+
+          if [[ $failure_count -eq $failure_limit ]]; then
+            >&2 echo "Failed to patch NodeUser. Number of attempts exceeded. NodeUser patch will be skipped."
+            patch_pending=false
+            break
+          fi
+
+          >&2 echo "Failed to patch NodeUser. ${failure_count} of ${failure_limit} attempts..."
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:89]:
```suggestion
          bb-log-error "Failed to patch NodeUser. ${failure_count} of ${failure_limit} attempts..."
```

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 9]
```diff
@@ -23,6 +23,105 @@ if bb-flag? disruption && bb-flag? reboot; then
...
+          sleep 10
+          continue
+        fi
+      done
+    done
+  else
+    >&2 echo "failed to patch NodeUser can't find kubelet.conf or bootstrap-token"
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:96]:
```suggestion
    bb-log-error "failed to patch NodeUser can't find kubelet.conf or bootstrap-token"
```

File: candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl [hunk 10]
```diff
@@ -61,85 +160,121 @@ function put_user_ssh_key() {
...
+        bb-log-error "Strange user with UID: $local_user_id, cannot delete it"
+        exit 0
+      fi
     fi
-  fi
-done
+  done
+}
+
+set +e
```
Comment (name212) [inline candi/bashible/common-steps/node-group/000_add_node_users.sh.tpl:276]: i think it needed for debugging purposes. Needs remove it

File: modules/040-node-manager/templates/bashible/rbac-for-us.yaml [hunk 11]
```diff
@@ -48,9 +48,11 @@ rules:
   - deckhouse.io
   resources:
   - instances/status
+  - nodeusers/status
```
Comment (name212) [inline modules/040-node-manager/templates/bashible/rbac-for-us.yaml:51]: Please create separated ClusterRole and ClusterRoleBinding for it

File: modules/040-node-manager/templates/bashible/rbac-for-us.yaml [hunk 12]
```diff
@@ -136,6 +138,9 @@ subjects:
   - kind: Group
     name: system:bootstrappers:d8-node-manager
     apiGroup: rbac.authorization.k8s.io
+  - kind: Group
```
Comment (name212) [inline modules/040-node-manager/templates/bashible/rbac-for-us.yaml:141]: https://github.com/deckhouse/deckhouse/pull/8054/files#r1577632019

File: modules/040-node-manager/templates/bashible/rbac-for-us.yaml [hunk 13]
```diff
@@ -48,9 +48,11 @@ rules:
   - deckhouse.io
   resources:
   - instances/status
+  - nodeusers/status
   verbs:
   - patch
   - update
+  - get
```
Comment (name212) [inline modules/040-node-manager/templates/bashible/rbac-for-us.yaml:55]: https://github.com/deckhouse/deckhouse/pull/8054/files#r1577632019

File: modules/040-node-manager/hooks/clear_nodeuser_errors.go [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+)
+
+type nodeUsersForClear struct {
+	Name         string
+	StatusErrors map[string]string
+}
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	Queue: "/modules/node-manager",
+	Kubernetes: []go_hook.KubernetesConfig{
```
Comment (name212) [inline modules/040-node-manager/hooks/clear_nodeuser_errors.go:45]: Looks like this hook should be schedule periodically as this hook https://github.com/deckhouse/deckhouse/blob/29d02b2f3fc8565e25142ff55241b25ed4e4cab0/modules/005-external-module-manager/hooks/cleanup_releases.go#L55

File: modules/040-node-manager/hooks/clear_nodeuser_errors.go [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+	return nodeUsersForClear{
+		Name:         nodeUser.Name,
+		StatusErrors: nodeUser.Status.Errors,
+	}, nil
+}
+
+func discoverNodeUsersForClear(input *go_hook.HookInput) error {
+	noSnap := input.Snapshots[nodeForClearSnapName]
+	nodes := make(map[string]struct{})
```
Comment (name212) [inline modules/040-node-manager/hooks/clear_nodeuser_errors.go:102]: can use go_lib.Set here

File: modules/040-node-manager/hooks/clear_nodeuser_errors.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+func hasIncorrectNodeUserErrors(nodeUserStatusError map[string]string, nodes map[string]struct{}) ([]string, bool) {
+	result := []string{}
+	has := false
+	for k := range nodeUserStatusError {
+		if _, ok := nodes[k]; !ok {
+			result = append(result, k)
+			has = true
+		}
+	}
+	return result, has
```
Comment (name212) [inline modules/040-node-manager/hooks/clear_nodeuser_errors.go:138]: we can return only result slice and check len of slice in the hook

File: modules/040-node-manager/hooks/clear_nodeuser_errors.go [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func hasIncorrectNodeUserErrors(nodeUserStatusError map[string]string, nodes map[string]struct{}) ([]string, bool) {
+	result := []string{}
```
Comment (name212) [inline modules/040-node-manager/hooks/clear_nodeuser_errors.go:130]:
```suggestion
	result := make([]string, 0, 0)
```

File: modules/040-node-manager/hooks/clear_nodeuser_errors.go
Comment (name212) [inline modules/040-node-manager/hooks/clear_nodeuser_errors.go:1]: Need test for this hook

**Review by name212 — CHANGES_REQUESTED** [review]
some nits

**Review by z9r5 — APPROVED** [review]

**Comment by aleksey-su** [comment]
image

**Comment by aleksey-su** [comment]
image

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @aleksey-su. I've started the workflow for label 'e2e/run/openstack'...

File: candi/bashible/bashbooster/65_nodeuser.sh [hunk 18]
```diff
@@ -0,0 +1,84 @@
+# Copyright 2024 Flant JSC
```
Comment (name212) [inline candi/bashible/bashbooster/65_nodeuser.sh:1]: move please to 000_add_node_users.sh.tpl

File: modules/040-node-manager/webhooks/validating/node_user [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+# limitations under the License.
+
+source /shell_lib.sh
+
+function __config__(){
+  cat <<EOF
+configVersion: v1
+kubernetes:
+  - name: nodeUsers
+    apiVersion: deckhouse.io/v1alpha1
```
Comment (name212) [inline modules/040-node-manager/webhooks/validating/node_user:24]: why deckhouse.io/v1alpha1 ?

Comment (name212) [inline modules/040-node-manager/webhooks/validating/node_user:24]:
```suggestion
    apiVersion: deckhouse.io/v1
```

**Review by name212 — APPROVED** [review]
two small things

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @aleksey-su. I've started the workflow for label 'e2e/run/openstack'...

**Review by RomanenkoDenys — APPROVED** [review]

**Pull request merged**